src/jit/morph.cpp

```diff
@@ -4354,32 +4476,113 @@ void Compiler::fgMorphMultiregStructArgs(GenTreeCall* call)
 //    For the GT_OBJ case will clone the address expression and generate two (or more)
 //    indirections.
 //    Currently the implementation only handles ARM64 and will NYI for other architectures.
-//    And for ARM64 we do not ye handle HFA arguments, so only 16-byte struct sizes are supported.
 //
 GenTreePtr    Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr fgEntryPtr)
 {
-    GenTreeArgList*  newArg = nullptr;
     assert(arg->TypeGet() == TYP_STRUCT);
-    GenTreePtr argValue = arg;
 
 #ifndef _TARGET_ARM64_
-    NYI("fgMorphMultiregStructArg non-ARM64 implementation");
+    NYI("fgMorphMultiregStructArg requires implementation for this target");
 #endif
 
-    // If we have a GT_OBJ of a GT_ADDR then
-    //  we set argValue to the child node ofthe GT_ADDR
+#if FEATURE_MULTIREG_ARGS
+    // Examine 'arg' and setup argClass 
+    //
+    CORINFO_CLASS_HANDLE objClass = nullptr;
+    GenTreePtr           argValue = arg;   // normally argValue will be arg, but see right below
     if (arg->OperGet() == GT_OBJ)
     {
-        GenTreePtr argAddr = arg->gtOp.gtOp1;
+        GenTreeObj*  argObj = arg->AsObj();
+        objClass = argObj->gtClass;
+
+        // If we have a GT_OBJ of a GT_ADDR then we set argValue to the child node of the GT_ADDR
+        //
+        if (argObj->gtOp1->OperGet() == GT_ADDR)
+        {
+            argValue = argObj->gtOp1->gtOp.gtOp1;
+        }
+    }
+    else if (arg->OperGet() == GT_LCL_VAR)
+    {
+        GenTreeLclVarCommon* varNode = arg->AsLclVarCommon();
+        unsigned varNum = varNode->gtLclNum;
+        assert(varNum < lvaCount);
+        objClass = lvaGetStruct(varNum);
+    }
+    noway_assert(objClass != nullptr);
 
-        if (argAddr->OperGet() == GT_ADDR)
+    unsigned   structSize = info.compCompHnd->getClassSize(objClass);
+    var_types  hfaType    = TYP_UNDEF;
+    var_types  elemType   = TYP_UNDEF;
+    unsigned   elemCount  = 0;
+    unsigned   elemSize   = 0;
+    var_types  type[MAX_MULTIREG_ARGS] = {};   // TYP_UNDEF = 0
+
+#ifdef FEATURE_HFA
+    hfaType = GetHfaType(objClass);   // set to float or double if it is an HFA, otherwise TYP_UNDEF
```

Comment:
The field isHfaRegArg tells us if we have an HFA passed in registers we also can have an HFA passed on the stack here.
And HFA are always passed by value on the stack where as other structs are passed by reference with the reference passed on the stack.